### PR TITLE
Support specifying `MaxBytes` in range request

### PR DESCRIPTION
**Pull request by linxiulei**
fixes: #14809

### CODE REVIEW

File: server/storage/mvcc/kvstore_txn.go
Comment (ahrtr) [inline server/storage/mvcc/kvstore_txn.go:133]: Do this check before `Unmarshal` (line 123)

File: server/storage/mvcc/kvstore_txn.go
Comment (ahrtr) [inline server/storage/mvcc/kvstore_txn.go:89]: Note you removed the `limit = limit + 1` in `txn.go`, so we should include the case `limit == len(revpairs)` as  well
```suggestion
	if limit <= 0 || limit >= len(revpairs) {
```

File: server/storage/mvcc/kvstore_txn.go
Comment (ahrtr) [inline server/storage/mvcc/kvstore_txn.go:137]: Why do yo copy the kvs?

**Comment by ahrtr** [comment]
The new feature makes sense to me,  only limiting the max number of entries isn't enough if each entry is too big.

When client specifies both limit and maxBytes, then server side truncates the result when whichever is satisfied first. 

It should have no impact on Kubernetes, because the maxByte defaults to 0, which means no limit.

**Comment by ahrtr** [comment]
Please also add some [common](https://github.com/etcd-io/etcd/tree/main/tests/common) e2e/integration test. Since this PR is already too big, so I'm OK to add it in a separate PR.

**Comment by ahrtr** [comment]
When this PR gets merged, probably someone could enhance K8s side to support `maxBytes` as well. cc @dims @liggitt @deads2k @lavalamp

FYI. https://github.com/etcd-io/etcd/pull/14810#issuecomment-1320984608

File: server/storage/mvcc/kvstore_txn.go
Comment (linxiulei) [reply server/storage/mvcc/kvstore_txn.go:137]: We need to shrink the slice to the actual length. The RangeResult returned will be used in caller. For example, the length will be used to determine if there is 'More'

File: server/storage/mvcc/kvstore_txn.go
Comment (linxiulei) [reply server/storage/mvcc/kvstore_txn.go]: Done

File: server/storage/mvcc/kv_test.go
Comment (ahrtr) [inline server/storage/mvcc/kv_test.go:312]: Please use golang subtest, there are lots of examples in the repo.  FYI. https://github.com/etcd-io/etcd/pull/14795

File: server/storage/mvcc/kvstore_txn.go
Comment (ahrtr) [reply server/storage/mvcc/kvstore_txn.go:138]: No need to copy the slice, you just need to generate a new slice based something like below,
```
kvs = kvs[:count]
```

Comment (linxiulei) [reply server/storage/mvcc/kvstore_txn.go]: Done. Thanks!

File: server/storage/mvcc/kv_test.go
Comment (linxiulei) [reply server/storage/mvcc/kv_test.go]: Done. Thanks!

**Review by ahrtr — APPROVED** [review]
LGTM

Thank yo @linxiulei 

cc @ptabor @serathius @spzala

**Comment by ahrtr** [comment]
@linxiulei Please also add a changelog in https://github.com/etcd-io/etcd/blob/main/CHANGELOG/CHANGELOG-3.6.md, of course, can be in a separate PR. 

TODO list:
1. add e2e/integration test. Please see https://github.com/etcd-io/etcd/pull/14810#issuecomment-1320985021;
2. add changelog item.

**Comment by linxiulei** [comment]
Will do. Thanks @ahrtr !

File: server/etcdserver/txn/txn.go
Comment (serathius) [inline server/etcdserver/txn/txn.go:201]: What's the reason to add this line?

File: server/etcdserver/txn/txn.go
Comment (linxiulei) [reply server/etcdserver/txn/txn.go:205]: I intentionally didn't implement this for maxBytes. It seems not a big change. Let me see how to it.

Comment (ahrtr) [reply server/etcdserver/txn/txn.go:205]: Actually I think we can even the following code, because `limit = limit + 1` was removed (line 136), and `len(rr.KVs) > int(r.Limit)` will never be true. But it isn't a big deal, so I did not raise the comment. 

```
if r.Limit > 0 && len(rr.KVs) > int(r.Limit) {
		rr.KVs = rr.KVs[:r.Limit]
		resp.More = true
	}
```

File: server/storage/mvcc/kvstore_txn.go
Comment (serathius) [inline server/storage/mvcc/kvstore_txn.go:131]: `count == i` Please remove redundancy

File: server/storage/mvcc/kvstore_txn.go
Comment (serathius) [inline server/storage/mvcc/kvstore_txn.go:132]: This is premature optimization. I expect that cost of calculating this on every loop is larger than saving we would get. 

Don't know what values of maxBytes you expect users to select, but let's say user sets 1MB. What's the probability that byte sizes align exactly to 1 MB?

File: server/etcdserver/txn/txn.go
Comment (serathius) [reply server/etcdserver/txn/txn.go:205]: Don't agree, please read 
https://github.com/etcd-io/etcd/blob/ddd65e7fe2e2d6fbb9a89a07ddbef79412885659/server/etcdserver/txn/txn.go#L127-L133

Condition `len(rr.KVs) > int(r.Limit)` can definitely be broken when sorting or filtering is enabled.

File: server/storage/mvcc/kvstore_txn.go
Comment (linxiulei) [reply server/storage/mvcc/kvstore_txn.go:131]: Not sure I understand this. We still need `count` outside of the loop to resize the slice at the end.

File: server/storage/mvcc/kvstore_txn.go
Comment (linxiulei) [reply server/storage/mvcc/kvstore_txn.go:132]: Do you want to combine this to the condition at line 122 (ie if `ro.MaxBytes > 0 && totalBytes >= ro.MaxBytes`)

Comment (linxiulei) [reply server/storage/mvcc/kvstore_txn.go:132]: This is not just optimization but also the correctness. We still need this KV pair if the total size equals to `maxBytes`

File: server/etcdserver/txn/txn.go
Comment (serathius) [inline server/etcdserver/txn/txn.go:222]:
```suggestion
		var totalBytes int64
		for i, kv := range rr.KVs {
			size := int64(kv.Size())
			if totalBytes + size > r.MaxBytes {
			    resp.More = true
				break
			}
            totalBytes += size
		}
```

File: server/storage/mvcc/kvstore_txn.go
Comment (serathius) [reply server/storage/mvcc/kvstore_txn.go:131]: Please see comment above about how similar for loop can be made simpler.

File: server/storage/mvcc/kvstore_txn.go
Comment (serathius) [reply server/storage/mvcc/kvstore_txn.go:132]: Please see comment above about how similar for loop can be made simpler.

**Comment by serathius** [comment]
I would like to see more tests before we merge this PR. This code is on a pretty important code path and there were already multiple edge cases not handled properly in this PR (for example maxBytes not working when sorting is enabled).

File: server/etcdserver/txn/txn.go
Comment (linxiulei) [reply server/etcdserver/txn/txn.go:205]: Done

File: server/storage/mvcc/kvstore_txn.go
Comment (linxiulei) [reply server/storage/mvcc/kvstore_txn.go]: Done

**Review by serathius — DISMISSED** [review]

File: server/storage/mvcc/kvstore_txn.go
Comment (fuweid) [inline server/storage/mvcc/kvstore_txn.go:131]: I think it should be `More: limit > len(kvs)`?

**Review by fuweid — APPROVED** [review]
LGTM(non-binding)

It seems that it reduces that grpc unary API memory usage. The grpc http2 transport only releases the memory after last byte has been sent. I was thinking that if the Range can be streaming type, the List-All from kube-apiserver won't cause memory spike in ETCD side because the pending bytes in memory is limit by http2 stream/connection windows size.  However, it might be hard to integrate.

This patch can chunk the data. Thanks! @linxiulei Help a lot.

**Comment by serathius** [comment]
Would also want to get some feedback from K8s scalability people who have some experience with issue that this feature tries to fix. cc @mborsz

**Comment by mborsz** [comment]
I think it would be worth discussing this with authors/reviewers of https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list which currently tries to solve similar (the same?) problem.


Probably it makes sense to get buy-in from sig-apimachinery that we will be able to use this in kube-apiserver before we make changes to etcd.

**Comment by linxiulei** [comment]
I agree to get more buy-ins from k8s side to make this feature more appealing. But I do not think it should be a blocker for etcd as this feature can benefit other use cases other than k8s. Also getting buy-ins from k8s side involves more discussions and takes time. 

Nonetheless, I will try to initiate some discussion within k8s.

File: server/etcdserver/txn/txn.go
Comment (ptabor) [inline server/etcdserver/txn/txn.go:212]: If I understand the code correctly, it means we might be sending '0' items, it is, we might stuck completely, 
giving empty responses and retries. 

I think we should prioritize giving at least 1 item in such situation. And be very explicit that in this situation the maxBytes contract can be exceeded.
But this is the exact semantic discussion we should have with k8s api machinery team @mborsz @lavalamp

We also need to have tests for this.

Comment (linxiulei) [reply server/etcdserver/txn/txn.go:212]: Yes, you understood correctly. The initial version was to return the result with size just exceeding the `maxBytes` so at least one item is returned. But this made the `maxBytes` a bit ambiguous. 

Given the max object size is 1.5MB, we can warm users the maxBytes should be bigger than 1.5MB to avoid returning 0 items. WDYT?

**Comment by lavalamp** [comment]
> When this PR gets merged, probably someone could enhance K8s side to support maxBytes as well

k8s would not directly expose this to end users, to be clear. (but thanks for the heads up!)

**Comment by linxiulei** [comment]
> Please also add some [common](https://github.com/etcd-io/etcd/tree/main/tests/common) e2e/integration test. Since this PR is already too big, so I'm OK to add it in a separate PR.

I added a commit of e2e tests in the previous push

**Comment by linxiulei** [comment]
How can I re-open it? I am still active on this PR if maintainers are able to evaluate it

**Comment by ahrtr** [comment]
I am not able to reopen either. Please raise a new PR, thx

**Comment by fuweid** [comment]
@linxiulei maybe it will be active after you repush?

**Comment by fuweid** [comment]
reopen...

**Comment by linxiulei** [comment]
repush didn't work either. Let me raise a new PR.

**Pull request closed without merging**
